### ISSUE TITLE
[RFE] Translate File Descriptors to File Names

### ISSUE THREAD

**Issue by mthbrown** [aquasecurity/tracee#1816]
## Prerequisites

- [ X] There isn't an issue describing the feature I need.
- [ X] I don't think opening a discussion thread first is relevant.
- [ X] I have a use case for the feature I would like to request.

## Feature description

This feature is about being able to easily determine what file some syscalls are related to. Currently, if I have tracee filter by something like the `write` syscall, I might see something like this:

```json
{
  "timestamp": 1654734049345486000,
  "processorId": 0,
  "processId": 7,
  "threadId": 7,
  "parentProcessId": 1,
  "hostProcessId": 233076,
  "hostThreadId": 233076,
  "hostParentProcessId": 233036,
  "userId": 0,
  "mountNamespace": 4026532349,
  "pidNamespace": 4026532352,
  "processName": "uname",
  "hostName": "55ab21515f70",
  "containerId": "55ab21515f707ca9a071f78b6efbac13d4d1ac9072fbf981a80989035ef943ee",
  "eventId": "1",
  "eventName": "write",
  "argsNum": 3,
  "returnValue": 6,
  "stackAddresses": null,
  "args": [
    {
      "name": "fd",
      "type": "int",
      "value": 1
    },
    {
      "name": "buf",
      "type": "void*",
      "value": 140337780654080
    },
    {
      "name": "count",
      "type": "size_t",
      "value": 6
    }
  ]
}
```

I have no idea what file this refers to and just have a file descriptor to work with. I'd have to try to go through previous syscalls to do the mapping. It would be great if there was an option to show the actual filename + path as well, similar to strace's `-y` option

The feature is useful because currently working with the output isn't easy

## Context

My current environment, where I would like the feature to work, is:

* Linux version: Ubuntu 20.04
* Linux kernel version: Linux ubuntu2004.localdomain 5.4.0-105-generic
* Tracee version (or commit id of your tree): Tracee version "v0.7.0"
* LLVM version:
* Golang version:

**Comment by geyslan** [comment]
https://github.com/aquasecurity/tracee/blob/260570df25844a7e05a1440d448082cec2447823/pkg/ebpf/events_pipeline.go#L161

[// Argument holds the information for one argument
type Argument struct {
	ArgMeta
	Value interface{} `json:"value"`
}](https://github.com/aquasecurity/tracee/blob/260570df25844a7e05a1440d448082cec2447823/types/trace/trace.go#L79-L83)

`Argument.Value` is an interface. For the fd case, is it ok to return it as a string `"1</etc/passwd>"` instead of an integer `1`?

**Comment by rafaeldtinoco** [comment]
Check the `--output option:parse-arguments` flag and what it does. I believe it will answer your question.

**Comment by geyslan** [comment]
I made a triage in tracee [events](https://github.com/aquasecurity/tracee/blob/260570df25844a7e05a1440d448082cec2447823/pkg/events/events.go#L191) and have found these syscalls which include `fd` between its args.

There are 8 special cases that the fd **isn't an int** and 3 others that **fd isn't the first syscall argument.** (see table below).

As I assume these syscalls are being [triggered](https://github.com/aquasecurity/tracee/blob/260570df25844a7e05a1440d448082cec2447823/pkg/ebpf/c/tracee.bpf.c#L2525) by `sys_enter` and it has a standard initialization `init_event_data()` for a bunch of structures like `event_data_t`, `task_info_t`, `syscall_data_t`, `args_t`, I would ponder if would be a good approach to put a **char[] buffer** inside `args_t` to be used in this case as the fd filepath: 

```c
typedef struct args {
    unsigned long args[6];
    char metadata[MAX_PATH_PREF_SIZE];
} args_t;
```

For the filepath retrieval, I think these helpers could be used in sequence:

https://github.com/aquasecurity/tracee/blob/260570df25844a7e05a1440d448082cec2447823/pkg/ebpf/c/tracee.bpf.c#L1142
https://github.com/aquasecurity/tracee/blob/260570df25844a7e05a1440d448082cec2447823/pkg/ebpf/c/tracee.bpf.c#L1137


```
-----------------------------------------------------
syscall             fd arg type   nth arg         set
-----------------------------------------------------
read                  int
write                 int
close                 int
fstat                 int
lseek                 int
mmap                  int           5            proc
ioctl                 int
pread64               int
pwrite64              int
readv                 int
writev                int
fcntl                 int
flock                 int
fsync                 int
fdatasync             int
ftruncate             int
getdents              int
fchdir                int
fchmod                int
fchown                int
fstatfs               int
readahead             int
fsetxattr             int
fgetxattr             int
flistxattr            int
fremovexattr          int
getdents64           uint
fadvise64             int
epoll_ctl             int           3
inotify_add_watch     int
inotify_rm_watch      int
sync_file_range       int
vmsplice              int
signalfd              int
fallocate             int
timerfd_settime       int                        time
timerfd_gettime       int                        time
signalfd4             int                     signals
preadv                int
pwritev               int
syncfs                int
setns                 int                        proc
finit_module          int                      system
preadv2               int
pwritev2              int
io_uring_enter       uint
io_uring_register    uint
faccessat2            int
epoll_pwait2          int
quotactl_fd          uint
readdir              uint                32bit_unique
llseek               uint                32bit_unique
mmap2               ulong           5    32bit_unique
ftruncate64           int                32bit_unique
fstat64               int                32bit_unique
fchown16             uint                32bit_unique
fcntl64               int                32bit_unique
fstatfs64             int                32bit_unique
fadvise64_64          int                32bit_unique
recvmmsg_time32       int                32bit_unique

```

**Comment by rafaeldtinoco** [comment]
Like we've already spoken, I believe having a map indexed by proc id (pid/tid) and fd #, with a path name, that could be accessed in userland if we want to enrich the file descriptor (parseArgs) BUT not tied to all events sent in the pipeline (due to sizing) could be a better approach I believe.

**Comment by rafaeldtinoco** [comment]
Let me know once you have a suggestion, or even better if you have a code proposal testing it if it would be enough or not. Thanks!

**Comment by geyslan** [comment]
I'll raise an initial PR as soon as possible. :+1:

**Comment by rafaeldtinoco** [comment]
This was solved/merged by: PR: #1971 / PR: #2086 

And quick pic showing the feature:

![image](https://user-images.githubusercontent.com/7395852/185836996-63bf1d30-3546-4b1e-a7e4-531a1a88a9b3.png)
